src/data/finset/basic.lean

```diff
@@ -2041,6 +2041,9 @@ iff.intro
 lemma attach_image_val [decidable_eq α] {s : finset α} : s.attach.image subtype.val = s :=
 eq_of_veq $ by rw [image_val, attach_val, multiset.attach_map_val, erase_dup_eq_self]
 
+@[simp] lemma attach_image_coe [decidable_eq α] {s : finset α} : s.attach.image coe = s :=
+finset.attach_image_val
```

Comment:
Does anything break if we add a simp lemma that says `subtype.val = coe`? If that existed you wouldn't need this lemma.

Reply:
I'm not sure if anything breaks (because I'd have to wait for a long build) but I don't quite see how `attach_image_coe` follows from:
```lean
@[simp] lemma subtype_val_eq_coe {α : Sort*} {p : α → Prop} :
  subtype.val = (coe : subtype p → α) :=
rfl
```
Are you sure this would suffice?

Reply:
You'd also need to mark the lemma above this one `simp`.

Reply:
Oh I understand now, but things do break if I try placing the lemma in what seemed like the right place.

To be more precise we need a slight reshuffle of what you have suggested: `attach_image_val` would have to be the lemma which follows `by simp` since the LHS of its statement contains `subtype.val`.

In other words, the following would work:
```lean
@[simp] lemma subtype_val_eq_coe {α : Sort*} {p : α → Prop} :
  subtype.val = (coe : subtype p → α) :=
rfl

@[simp] lemma attach_image_coe [decidable_eq α] {s : finset α} : s.attach.image coe = s :=
eq_of_veq $ by rw [← subtype_val_eq_coe, image_val, attach_val, multiset.attach_map_val, erase_dup_eq_self]

lemma attach_image_val [decidable_eq α] {s : finset α} : s.attach.image subtype.val = s :=
by simp only [attach_image_coe, subtype_val_eq_coe]
```

So I tried placing `subtype_val_eq_coe` in core [here](https://github.com/leanprover-community/lean/blob/bef30f3ea338cf0e068d942649a613eabf162920/library/init/coe.lean#L161) however this immediately triggered breakage [here](https://github.com/leanprover-community/lean/blob/bef30f3ea338cf0e068d942649a613eabf162920/library/init/data/fin/ops.lean#L99) (replacing `dsimp` with `dsimp only` fixes it) so I'm wary of fighting for this win.

Reply:
I think a safer place would be next to

https://github.com/leanprover-community/mathlib/blob/4b5bf56bd3ca987620453f27491737c8eedc68a7/src/data/subtype.lean#L39

For the sake of this PR, how about you keep your new lemma and mark the `val` lemma above as simp too?

Reply:
I've just pushed https://github.com/leanprover-community/mathlib/pull/9103/commits/e2e4389d46921c4a4283615449d7b8c99b6bb7ec to see if things build with `subtype_val_eq_coe` added where you have suggested.

The build didn't break after 6 minutes (562% CPU) on my machine so there's _some_ change it might work. If this fails, I'll eagerly adopt your compromise.

Reply:
Aaaaand that failed [here](https://github.com/leanprover-community/mathlib/runs/3730275807#step:6:11).

I've gone with the compromise of just adding `simp` to `attach_image_val`. If https://github.com/leanprover-community/mathlib/pull/9103/commits/3a4d551d06694d70abfef527182a125a62833967 builds successfully, I'll resolve this conversation.

Reply:
https://leanprover-community.github.io/mathlib_docs/data/subtype.html#subtype.val_eq_coe already exists, right?

Reply:
> https://leanprover-community.github.io/mathlib_docs/data/subtype.html#subtype.val_eq_coe already exists, right?

Yes. I guess Eric's point is that it should probably state:
```lean
∀ {α : Sort*} {p : α → Prop}, val = (coe : subtype p → α)
```
rather than
```lean
∀ {α : Sort*} {p : α → Prop} {x : subtype p}, x.val = (x : α)
```
as it currently does.

However the change breaks things and I'd need a bit more encouragement to fold this change into this PR.